### DISCUSSION TITLE
ImportException when importing built-in libraries in Windows Service

### DISCUSSION THREAD

**Discussion by JBensonNBT** [IronLanguages/ironpython3#1668] [General]
I'm building a Windows Service using IronPython 3.4 to process scripts. I have discovered strange behavior when running scripts in my service that don't occur in a standard Windows console app.

The following examples *always* generate an ImportException in the service:
Python line:  "import time"
Python line: "from time import *"
IronPython C# (scope = ScriptScope): "scope.ImportModule("time")

Similar exceptions occur for other built-in modules (e.g. math).

The same examples execute properly in standard Windows running under my user account. My Windows service is running under the Local Service account. The runtime environment seems to be the major difference between success and failure.

I am new to IrionPython so it's certainly possible I'm doing something wrong settiing things up:
ScriptEngine pythonEngine = Python.CreateEngine();
ScriptSource pythonScript = pythonEngine.CreateScriptSourceFromString(script.Script, SourceCodeKind.AutoDetect);
pythonScript.Execute(scope);

ImportException always occurs on the Execute(scope) step if the Python script includes an "import" either specifically in my code or inside a .py file imported from the standard libarary.

Here is an example of the Windows exception stack:
[Ironpython_time_import_exception.txt](https://github.com/IronLanguages/ironpython3/files/10936288/Ironpython_time_import_exception.txt)

This is turning out to be a showstopper. Anyone have ideas how to overcome the problem?  IronPython bug maybe?
Thanks
Jeff

**Comment by slozier** [comment]
It sounds like it might be failing to load the `IronPython.Modules.dll`? Are you able to import the built-in `sys` modules (which is implemented directly in `IronPython.dll`).

**Reply by JBensonNBT** [reply]
Thanks for the quick reply. You just might be on to something. I did notice that "import sys" worked whereas "import math" or "import time" did not. Looking at the build output executables, I don't see IronPython.Modules.dll. By comparison, my test app has a variety of additional IronPython DLLs, including the one you mention.

The architecture of my service is all the code related to Python is in a separate DLL from the main service app. The DLL project contains all the expected IronPython references. It seems that when the complete service builds only the main IronPython.dll gets copied to the bin folder. Any ideas how to get those other IronPython dependencies to be copied by the Visual Studio compilation?

Thanks for the help
-J

**Reply by slozier** [reply]
I assume it's being pruned because `IronPython.Modules.dll` is not directly referenced by `IronPython.dll` so VS thinks you don't need it. There's probably a better solution, but you could "make use" of it in your DLL. Could be anything, but I tried with something like the following and it seems to work:
```cs
private static Type referenceIronPythonModules  = typeof(IronPython.Modules.PythonTime);
```

**Reply by JBensonNBT** [reply]
Thanks again. I have confirmed that by copying the missing IronPython DLLs to my binary output folder "by hand" solves the import exception problem in my service. I will wrestle with the VS issues separately. In some ways this might be a small advantage. Because of the nature of our application we might not want to enable all the features of IronPython for all users.
